app/models/project_topic.rb

```diff
@@ -0,0 +1,95 @@
+  end
+
+  # Adds a new sign-up to the waitlist by updating the sign-up record.
+  # Marks the sign-up as waitlisted and associates it with the current topic.
+  def save_waitlist_entry(new_sign_up)
+    new_sign_up.update(is_waitlisted: true, sign_up_topic_id: self.id)
+  end
+
+  # Signs up a team for the current topic.
+  # Checks if the team is already signed up, and if so, ensures they are not waitlisted.
+  # If a slot is available, assigns the topic to the team; otherwise, adds the team to the waitlist.
+  def sign_up_team(team_id)
+    topic_id = self.id
+
+    # Check if the team has already signed up for this topic
+    existing_sign_up = SignedUpTeam.find_by(sign_up_topic_id: topic_id, team_id: team_id)
+
+    # If the team is already signed up and not waitlisted, return false
+    if !existing_sign_up.nil? && !existing_sign_up.is_waitlisted
+      return false
+    end
+
+    # Create a new sign-up entry for the team
+    new_sign_up = SignedUpTeam.new(sign_up_topic_id: topic_id, team_id: team_id)
+
+     # If there are available slots, assign the topic to the team and remove the team from the waitlist
+    if slot_available?
+      assign_topic_to_team(new_sign_up)
+      result = SignedUpTeam.drop_off_team_waitlists(team_id)
+    else
+      # If no slots are available, add the team to the waitlist
+      result = save_waitlist_entry(new_sign_up)
+    end
+
+    result
+  end
+
+  # Retrieves the team that has been waitlisted the longest for a given topic.
+  # The team is selected based on the earliest created waitlist entry.
+  def longest_waiting_team
+    SignedUpTeam.where(sign_up_topic_id: self.id, is_waitlisted: true).order(:created_at).first
+  end
+
+  # Removes a team from the current topic.
+  # If the team is not waitlisted, the next waitlisted team is reassigned to the topic.
+  # The team is then destroyed (removed from the sign-up record).
+  def drop_team_from_topic(team_id)
+    # Find the sign-up record for the team for this topic
+    signed_up_team = SignedUpTeam.find_by(team_id: team_id, sign_up_topic_id: self.id)
+    return nil unless signed_up_team
+
+    # If the team is not waitlisted, reassign the topic to the next waitlisted team
+    unless signed_up_team.is_waitlisted
+      next_waitlisted_team = longest_waiting_team
+      next_waitlisted_team&.reassign_topic(self.id)
+    end
+    
+    # Destroy the sign-up record for the team
+    signed_up_team.destroy
+  end
+
+  # Retrieves all teams that are signed up for a given topic.
+  def signed_up_teams_for_topic
+    SignedUpTeam.where(sign_up_topic_id: self.id)
+  end
+
+  # Calculates the number of available slots for a topic.
+  # It checks how many teams have already chosen the topic and subtracts that from the maximum allowed choosers.
+  def current_available_slots
+    # Find the teams who have already chosen the topic and are not waitlisted
+    teams_who_chose_the_topic = SignedUpTeam.where(sign_up_topic_id: self.id, is_waitlisted: false)
```

Comment:
teams_who_chose_the_topic seems to imply that it includes waitlisted teams, but it doesn't.  Please find a better name.  Also, I think the code would be more readable if you apply the .size operator in Line 90 instead of Line 93.